From 45d580da449871f769556522be5a0f3409dc9b03 Mon Sep 17 00:00:00 2001
From: Dimitri John Ledkov <dimitri.ledkov@surgut.co.uk>
Date: Tue, 1 Jul 2025 11:51:56 +0100
Subject: [PATCH] install: add support for portable libcrypto

Add support in install command to also copy libcrypto and associated
files to run on the host with container provided OpenSSL.
---
 cni/pkg/install/binaries.go | 37 +++++++++++++++++++++++++++++++++++++
 1 file changed, 37 insertions(+)

diff --git a/cni/pkg/install/binaries.go b/cni/pkg/install/binaries.go
index 5d78c30714..7a461d55e8 100644
--- a/cni/pkg/install/binaries.go
+++ b/cni/pkg/install/binaries.go
@@ -26,6 +26,43 @@ import (
 // and returns a set of the filenames copied.
 func copyBinaries(srcDir string, targetDirs []string) (sets.String, error) {
 	copiedFilenames := sets.String{}
+
+	cgr_files := []string{
+		"/etc/ssl/fipsmodule.cnf",
+		"/etc/ssl/openssl.cnf",
+		"/usr/lib/libcrypto3/libcrypto.so.3",
+		"/usr/lib/ossl-modules/fips.so",
+	}
+	for _, srcFilepath := range cgr_files {
+		filename := filepath.Base(srcFilepath)
+
+		for _, targetDir := range targetDirs {
+			if err := os.MkdirAll(filepath.Join(targetDir, ".cgr"), 0o755); err != nil {
+				installLog.Errorf("failed to create .cgr/ in %s: %s", targetDir, err.Error())
+			}
+
+			// remove previous tmp file if some exist before creating a new one
+			// the only possible returned error is [ErrBadPattern], when pattern is malformed. Can be ignored in this case
+			matches, _ := filepath.Glob(filepath.Join(targetDir, filename) + ".tmp.*")
+			if len(matches) > 0 {
+				installLog.Infof("Target folder %s contains one or more temporary files with a %s name. The temp files will be deleted.", targetDir, filename)
+				for _, file := range matches {
+					if err := os.Remove(file); err != nil {
+						installLog.Warnf("Failed to delete tmp file %s from previous run: %s", file, err)
+					}
+				}
+			}
+
+			if err := file.AtomicCopy(srcFilepath, targetDir, filename); err != nil {
+				installLog.Errorf("failed file copy of %s to %s: %s", srcFilepath, targetDir, err.Error())
+				return copiedFilenames, err
+			}
+			installLog.Infof("Copied %s to %s", filename, targetDir)
+
+		}
+		copiedFilenames.Insert(filename)
+	}
+
 	srcFiles, err := os.ReadDir(srcDir)
 	if err != nil {
 		return copiedFilenames, err
-- 
2.48.1

